src/language_features/document_symbol.rs

```diff
@@ -332,25 +332,43 @@ fn next_or_prev_symbol_details<T: Symbol<T> + 'static>(
         let symbol_position =
             get_kakoune_position_with_fallback(&meta.buffile, symbol_position, ctx);
 
-        if exceeds(
-            symbol_position,
-            kind,
-            params.position,
-            params.search_next,
-            &params.symbol_kind,
-        ) {
-            return Some((
-                filename.to_string(),
+        let symbol_name = symbol.name().to_owned();
+
+        // Order of checking the parent symbol for exceeds() matters.
+        // If we're searching forward then check if the parent symbol exceeds() the current
+        // symbol first. If we're searching backwards then check if any of children symbols
+        // exceeds() the current symbol first
+
+        // Deals with the case we're searching forwards
+        if params.search_next
+            && exceeds(
                 symbol_position,
-                symbol.name().to_owned(),
                 kind,
-            ));
+                params.position,
+                params.search_next,
+                &params.symbol_kind,
+            )
+        {
+            return Some((filename, symbol_position, symbol_name, kind));
         }
 
         let from_children = next_or_prev_symbol_details(symbol.children(), params, meta, ctx);
         if from_children.is_some() {
             return from_children;
         }
+
+        // Deals with the case we're searching backwards
+        if !params.search_next
+            && exceeds(
```

Comment:
you now duplicate the lengthy call to exceeds().
My first idea was to compute it in one place and share that.

```rust
let beyond_cursor = if exceeds(...) {
    Some((filename, symbol_position, symbol_name, kind))
} else {
    None
};

if params.search_next && beyond_cursor.is_some() {
    return beyond_cursor;
}

if let Some(child_beyond_cursor) =
    next_or_prev_symbol_details(symbol.children(), params, symbol_kinds, meta, ctx)
{
    return Some(child_beyond_cursor);
}

if !params.search_next && beyond_cursor.is_some() {
    return beyond_cursor;
}
```

but actually, it seems better to just remove exceeds() altogether.
Beyond checking the symbol kind (which is not obvious from its name), it does exactly 

```rust
    if search_next {
        symbol_position > cursor_position
    } else {
        symbol_position < cursor_position
    }
```

which means it definitely no longer pulls its weight if the callers *already* branch on `search_next`.
So the whole thing should be simplified to something like

```rust
let want_symbol = symbol_kind.is_empty() || format!("{:?}", kind) == symbol_kind;

if params.search_next && want_symbol && symbol_position > cursor_position {
    return Some((filename, symbol_position, symbol_name, kind));
}

if let Some(from_children) =
    next_or_prev_symbol_details(symbol.children(), params, symbol_kinds, meta, ctx)
{
    return Some(from_children);
}

if params.search_next && want_symbol && symbol_position < cursor_position {
    return Some((filename, symbol_position, symbol_name, kind));
}
```

(having removed this overzealous abstraction, I also see it as strange that we use < and > instead of something total like  < and >= but I haven't found problems)

Reply:
Removal of exceeds sound great!